inst/include/lpdens.hpp

```diff
@@ -5,231 +5,286 @@
-                                                  x_min - 0.05 * range);
-            grid_size -= 5;
+    } else {
+        if (!std::isnan(xmin_) & !std::isnan(xmax_)) {
+            // two boundaries -> probit transform
+            x_new = stats::pnorm(x).array() + xmin_ - 5e-3;
+            x_new *=  (xmax_ - xmin_ + 1e-2);
+        } else if (!std::isnan(xmin_)) {
+            // left boundary -> log transform
+            x_new = x.array().exp() + xmin_ - 1e-3;
+        } else if (!std::isnan(xmax_)) {
+            // right boundary -> negative log transform
+            x_new = -(x.array().exp() - xmax_ - 1e-3);
         } else {
-            lowr_ext = Eigen::VectorXd();
+            // no boundary -> no transform
         }
+    }
 
-        // no right boundary -> add a few points to the right
-        if (std::isnan(xmax_)) {
-            uppr_ext = Eigen::VectorXd::LinSpaced(5,
-                                                  x_max + 0.05 * range,
-                                                  x_max + 0.5 * range);
-            grid_size -= 5;
-        } else {
-            uppr_ext = Eigen::VectorXd();
-        }
-
-        grid_points <<
-            lowr_ext,
-            Eigen::VectorXd::LinSpaced(grid_size, x_min, x_max),
-            uppr_ext;
-
-        return grid_points;
+    return x_new;
+}
+
+//! corrects the density estimate for a preceding boundary transformation of
+//! the data.
+//! @param x evaluation points (in original domain).
+//! @param fhat the density estimate evaluated in the transformed domain.
+//! @return corrected density estimates at `x`.
+inline Eigen::VectorXd LPDens1d::boundary_correct(const Eigen::VectorXd& x,
+                                                  const Eigen::VectorXd& fhat)
+{
+    Eigen::VectorXd corr_term(fhat.size());
+    if (!std::isnan(xmin_) & !std::isnan(xmax_)) {
+        // two boundaries -> probit transform
+        corr_term = (x.array() - xmin_ + 5e-3) / (xmax_ - xmin_ + 1e-2);
+        corr_term = stats::dnorm(stats::qnorm(corr_term));
+        corr_term /= (xmax_ - xmin_ + 1e-2);
+        corr_term = 1 / corr_term.array().max(1e-4);
+    } else if (!std::isnan(xmin_)) {
+        // left boundary -> log transform
+        corr_term = 1 / (1e-3 + x.array() - xmin_);
+    } else if (!std::isnan(xmax_)) {
+        // right boundary -> negative log transform
+        corr_term = 1 / (1e-3 + xmax_ - x.array());
+    } else {
+        // no boundary -> no transform
+        corr_term.fill(1.0);
     }
 
-    Eigen::VectorXd finalize_grid(Eigen::VectorXd& grid_points)
-    {
-        if (!std::isnan(xmin_))
-            grid_points(0) = xmin_;
-        if (!std::isnan(xmax_))
-            grid_points(grid_points.size() - 1) = xmax_;
-        return grid_points;
+    return fhat.array() * corr_term.array();
+}
+
+//! constructs a grid that is later used for interpolation.
+//! @param x vector of observations.
+//! @return a grid of size 200.
```

Comment:
50, and shouldn't it be an argument (set by default to 50)?

Reply:
fixed the typo, would not make it argument until we think it's useful.